src/GitHub.TeamFoundation.14/Services/VSGitExt.cs

```diff
@@ -1,39 +1,109 @@
+            ActiveRepositories = new ILocalRepositoryModel[0];
+
+            if (context.IsActive && TryInitialize())
+            {
+                // Refresh ActiveRepositories on background thread so we don't delay startup.
+                InitializeTask = Task.Run(() => RefreshActiveRepositories());
+            }
+            else
+            {
+                // If we're not in the UIContext or TryInitialize fails, have another go when the UIContext changes.
+                context.UIContextChanged += ContextChanged;
+                log.Information("VSGitExt will be initialized later");
+                InitializeTask = Task.CompletedTask;
+            }
         }
 
+        void ContextChanged(object sender, VSUIContextChangedEventArgs e)
+        {
+            // If we're in the UIContext and TryInitialize succeeds, we can stop listening for events.
+            // NOTE: this event can fire with UIContext=true in a TFS solution (not just Git).
+            if (e.Activated && TryInitialize())
+            {
+                // Refresh ActiveRepositories on background thread so we don't delay UI context change.
+                InitializeTask = Task.Run(() => RefreshActiveRepositories());
+                context.UIContextChanged -= ContextChanged;
+                log.Information("Initialized VSGitExt on UIContextChanged");
+            }
+        }
 
-        void CheckAndUpdate(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        bool TryInitialize()
         {
-            Guard.ArgumentNotNull(e, nameof(e));
-            if (e.PropertyName != "ActiveRepositories" || gitService == null)
-                return;
-            ActiveRepositoriesChanged?.Invoke();
+            gitService = serviceProvider.GetService<IGitExt>();
+            if (gitService != null)
+            {
+                gitService.PropertyChanged += (s, e) =>
+                {
+                    if (e.PropertyName == nameof(gitService.ActiveRepositories))
+                    {
+                        RefreshActiveRepositories();
+                    }
+                };
+
+                log.Information("Found IGitExt service and initialized VSGitExt");
+                return true;
+            }
+
+            log.Information("Couldn't find IGitExt service");
+            return false;
         }
 
-        public IEnumerable<ILocalRepositoryModel> ActiveRepositories => gitService?.ActiveRepositories.Select(x => x.ToModel());
+        void RefreshActiveRepositories()
+        {
+            try
+            {
+                ActiveRepositories = gitService?.ActiveRepositories.Select(x => repositoryFactory.Create(x.RepositoryPath)).ToList();
+                ActiveRepositoriesChanged?.Invoke();
+            }
+            catch (Exception e)
+            {
+                log.Error(e, "Error refreshing repositories");
+                ActiveRepositories = new ILocalRepositoryModel[0];
```

Comment:
Shouldn't this be raising the `ActiveRepositoriesChanged` event as well? And if so, shouldn't this event be raised in the setter of the property directly?

Reply:
Oops! Good catch. This is why I should have done it on the property itself!